module/zfs/spa.c

```diff
@@ -4513,6 +4513,214 @@ spa_vdev_add(spa_t *spa, nvlist_t *nvroot)
 	return (0);
 }
 
+typedef struct {
+	vdev_t  *ssa_newvd;
+	vdev_t  *ssa_draid;
+	uint64_t ssa_dtl_max;
+} sequential_scan_arg_t;
+
+int vdev_draid_max_rebuild = 128;
+
+static void
+vdev_draid_scan_done(zio_t *zio)
+{
+	spa_t *spa = zio->io_spa;
+	dsl_scan_t *scn = spa->spa_dsl_pool->dp_scan;
+
+	ASSERT(zio->io_bp != NULL);
+
+	zio_data_buf_free(zio->io_data, zio->io_size);
+	kmem_free(zio->io_private, sizeof(blkptr_t));
+
+	scn->scn_phys.scn_examined += DVA_GET_ASIZE(&zio->io_bp->blk_dva[0]);
+	spa->spa_scan_pass_exam += DVA_GET_ASIZE(&zio->io_bp->blk_dva[0]);
+
+	mutex_enter(&spa->spa_scrub_lock);
+
+	spa->spa_scrub_inflight--;
+	cv_broadcast(&spa->spa_scrub_io_cv);
+
+	if (zio->io_error && (zio->io_error != ECKSUM ||
+	    !(zio->io_flags & ZIO_FLAG_SPECULATIVE))) {
+		spa->spa_dsl_pool->dp_scan->scn_phys.scn_errors++;
+	}
+
+	mutex_exit(&spa->spa_scrub_lock);
+}
+
+static void
+vdev_draid_rebuild_block(zio_t *pio, vdev_t *vd, uint64_t offset, uint64_t asize)
+{
+	/* HH: maybe bp can be on the stack */
+	blkptr_t *bp = kmem_alloc(sizeof(*bp), KM_SLEEP);
+	dva_t *dva = bp->blk_dva;
+	uint64_t psize;
+	spa_t *spa = vd->vdev_spa;
+
+	ASSERT3P(vd->vdev_ops, ==, &vdev_mirror_ops);
+
+	psize = asize;
+	ASSERT3U(asize, ==, vdev_psize_to_asize(vd, psize));
+
+	mutex_enter(&spa->spa_scrub_lock);
+	while (spa->spa_scrub_inflight > vdev_draid_max_rebuild)
+		cv_wait(&spa->spa_scrub_io_cv, &spa->spa_scrub_lock);
+	spa->spa_scrub_inflight++;
+	mutex_exit(&spa->spa_scrub_lock);
+
+	BP_ZERO(bp);
+
+	DVA_SET_VDEV(&dva[0], vd->vdev_id);
+	DVA_SET_OFFSET(&dva[0], offset);
+	DVA_SET_GANG(&dva[0], !!0);
```

Comment:
Yes should be just 0. The !!0 was there because the code was originally _DVA_SET_GANG(&dva[0], !!(flags & ZDB_FLAG_GBH));_